packages/lit-html/src/disconnectable-directive.ts

```diff
@@ -0,0 +1,384 @@
+ * recursively run `setConnected()` over the entire tree, without removing any
+ * children from `_$disconnectableChildren`, since this tree is required to
+ * re-connect the tree, which does the same operation, simply passing
+ * `isConnectd: true` down the tree, signaling which callback to run.
+ */
+
+import {
+  Directive,
+  PartInfo,
+  NodePart,
+  NODE_PART,
+  DisconnectableParent,
+  noChange,
+} from './lit-html.js';
+import {setPartValue} from './parts.js';
+export {directive} from './lit-html.js';
+
+/**
+ * Recursively walks down the tree of Parts/TemplateInstances/Directives to set
+ * the connected state of directives and run `disconnectedCallback`/
+ * `reconnectedCallback`s.
+ */
+const setChildrenConnected = (
+  parent: DisconnectableParent,
+  isConnected: boolean
+): boolean => {
+  const children = parent._$disconnetableChildren;
+  if (children !== undefined) {
+    for (const child of children) {
+      // The existence of `_$setDirectiveConnected` is used as a "brand" to
+      // disambiguate DisconnectableDirectives from other DisconnectableChildren
+      // (as opposed to using an instanceof check to know when to call it); the
+      // redundancy of "Directive" in the API name is to avoid conflicting with
+      // `_$setNodePartConnected`, which exists `NodeParts` which are also in
+      // this list
+      if ((child as Directive)._$setDirectiveConnected) {
+        // Disconnect Directive (and any nested directives contained within)
+        (child as DisconnectableDirective)._$setDirectiveConnected(
+          isConnected,
+          false
+        );
+      } else {
+        // Disconnect Part/TemplateInstance
+        setChildrenConnected(child, isConnected);
+      }
+    }
+    return true;
+  }
+  return false;
+};
+
+/**
+ * Removes the given child from its parent list of disconnectable children, and
+ * if the parent list becomes empty as a result, removes the parent from its
+ * parent, and so forth up the tree when that causes subsequent parent lists to
+ * become empty.
+ */
+const removeFromParent = (child: DisconnectableParent) => {
+  let parent = child._$parent;
+  let children = parent?._$disconnetableChildren;
+  children?.delete(child);
+  while (children !== undefined && children.size === 0) {
+    child._$disconnetableChildren = undefined;
+    if (parent !== undefined) {
+      children = parent._$disconnetableChildren!;
+      children.delete(child);
+    } else {
+      break;
+    }
+    child = parent;
+    parent = child._$parent;
+  }
+};
+
+/**
+ * Sets the connected state on any directives contained within the committed
+ * value of this part (i.e. within a TemplateInstance or iterable of NodeParts)
+ * and runs their `disconnectedCallback`/`reconnectedCallback`s, as well as
+ * within any directives stored on the NodePart (when `valueOnly` is false).
+ *
+ * `isClearingValue` should be passed as `true` on a top-level part that is
```

Comment:
Seems like this parameter is similar to `shouldRemoveFromParent` in `_$setDirectiveConnected` in that both are true at the top-level and false otherwise. Could they be named similarly?

Reply:
Done.
In `_$setNodePartConnected` it's `isClearingValue` and 
in `_$setDirectiveConnected` it's now `isClearingDirective`.